### PR TITLE
Feature/event listener

### PR DESCRIPTION
**Pull request by rakeshsahuu**
Added support of Event Listener service by introducing the following command,
(redfish) help listener service

 ==========================================================================================================================================================================

 Command: listener service <start | stop>

 Description:
 
 This command will start/ stop the Redfish event listener (using the specified IP and port via http/ https), based on the configuration
 
 'listener service start'   - starts the event listener
 'listener service stop'    - stops the event listener
 
 Notes,
 1. Add this listener destination subscription in the system using below command, to see the events reaching to this listener service,
    'http post /redfish/v1/EventService/Subscriptions json/EventDestination.json'
 2. Show this listener destination subscription of the system,
    'http get /redfish/v1/EventService/Subscriptions/<listener destination subscriber "Id">'
 3. Delete this listener destination subscription from the system using below command,
    'http delete /redfish/v1/EventService/Subscriptions/<listener destination subscriber "Id">'
 4. Check if Test event reaching to the listener service,
    'http post /redfish/v1/EventService/Actions/EventService.SubmitTestEvent json/testEvent.json'
    then, look in "Events_<system IP>.txt" for the Test event
 
 Example:
 
 (redfish) listener service start
 Starting the listener service on 10.237.117.212:8080 via HTTP
 
 (redfish)
 
 ==========================================================================================================================================================================

Internally, on  'listener service start' command, the listener service is started as a separate thread. And each received event packet is processed in a separate thread.

[Event Listener Testing Logs](https://github.com/Seagate/SystemsRedfishPy/files/9971468/EventListenerTestLog.txt)

### CODE REVIEW

File: redfishAPI.cfg [hunk 1]
```diff
@@ -17,5 +17,10 @@
     "trace": "4",
     "urltimeout": "30",
     "usefinalslash": "True",
-    "username": ""
+    "username": "",
+    "listenerusessl": "False",
+    "listenerip": "localhost",
+    "listenerport": "8080",
+    "certfile": "cert.pem",
+    "keyfile": "server.key"
```
Comment (jskazinski) [inline redfishAPI.cfg:25]: The certificate PEM file and private key PEM file should not be hard-coded in the configuration file. The user should specify the security files required.

File: server.key [hunk 2]
```diff
@@ -0,0 +1,51 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (jskazinski) [inline server.key:1]: This file should not be added to the repo.

File: cert.pem [hunk 3]
```diff
@@ -0,0 +1,29 @@
+-----BEGIN CERTIFICATE-----
```
Comment (jskazinski) [inline cert.pem:1]: This file should not be added to the repo.

File: core/redfishConfig.py [hunk 4]
```diff
@@ -67,6 +68,11 @@ def __init__(self, filename):
         self.dictionary['urltimeout']       = [300, '<int>       How long to wait for a URL request before timing out. Default is 300.']
         self.dictionary['usefinalslash']    = [True, 'True|False  When True, all Redfish URIs will have a slash as the final character in the URL. Default is True.']
         self.dictionary['username']         = ['', '<string>    Change the username to [name] that is used to log in to the Redfish Service.']
+        self.dictionary['listenerusessl']   = ["False", 'True|False  Switch between use http and https. Default is https.']
```
Comment (jskazinski) [inline core/redfishConfig.py:71]:
```suggestion
        self.dictionary['listenerusessl']   = ["True", 'True|False  Switch between use http and https. Default is https.']
```

File: core/redfishConfig.py [hunk 5]
```diff
@@ -67,6 +68,11 @@ def __init__(self, filename):
...
+        self.dictionary['listenerip']       = ['localhost', '<string>    Event and Telemetry Listener IP address.']
+        self.dictionary['listenerport']     = ['8080', '<string>    Event and Telemetry Listener port.']
+        self.dictionary['certfile']         = ['cert.pem', '<string>    Certificate for ssl connection.']
+        self.dictionary['keyfile']          = ['server.key', '<string>    Key for ssl connection.']
```
Comment (jskazinski) [inline core/redfishConfig.py:75]:
```suggestion
        self.dictionary['certfile']         = ['', '<string>    Certificate PEM file for the SSL connection.']
        self.dictionary['keyfile']          = ['', '<string>    Private Key PEM file for the SSL connection.']
```

File: commands/common/listener_service.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
+#
+# Do NOT modify or remove this copyright and license
+#
+# Copyright (c) 2019 Seagate Technology LLC and/or its Affiliates, All Rights Reserved
```
Comment (jskazinski) [inline commands/common/listener_service.py:4]:
```suggestion
# Copyright (c) 2022 Seagate Technology LLC and/or its Affiliates, All Rights Reserved
```

**Review by jskazinski — CHANGES_REQUESTED** [review]
Rakesh, really excited to see you working on this and seeing your solution. Thank you for the contribution. I had some general comments to address before merging. Please all update version.py to '2.5.0' to indicate this significant feature. Please also update CHANGELOG.md.

File: redfishAPI.cfg (hunk 1 above)
Comment (rakeshsahuu) [reply redfishAPI.cfg:25]: Done.

File: server.key (hunk 2 above)
Comment (rakeshsahuu) [reply server.key:1]: Done.

File: cert.pem (hunk 3 above)
Comment (rakeshsahuu) [reply cert.pem:1]: Done.

File: core/redfishConfig.py (hunk 4 above)
Comment (rakeshsahuu) [reply core/redfishConfig.py:71]: Done. I missed that.

File: commands/common/listener_service.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+                    redfishConfig.listener.start()
+                except Exception as err:
+                    Trace.log(TraceLevel.ERROR, traceback.print_exc())
+            else:
+                Trace.log(TraceLevel.INFO, 'Listener service already running')
+        elif self.subcommand == 'stop':
+            redfishConfig.listener.stop = True
+            redfishConfig.listener.join()
+            redfishConfig.listener = None
+            Trace.log(TraceLevel.INFO, 'Listener service successfully stopped')
```
[the remark this answers is not in the document]
Comment (rakeshsahuu) [reply commands/common/listener_service.py:92]: Done.

File: redfishAPI.py [hunk 8]
```diff
@@ -73,6 +73,12 @@
         # Run script mode
         returncode = RedfishScript.execute_script(redfishConfig, args.scriptfile)
 
+    # Before existing, stop the listener service if running
+    if redfishConfig.listener != None:
+        redfishConfig.listener.stop = True
+        redfishConfig.listener.join()
+        redfishConfig.listener = None
```
[the remark this answers is not in the document]
Comment (rakeshsahuu) [reply redfishAPI.py:80]: Done.

File: commands/common/listener_service.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+                if p.method() == 'GET':
+                    res = "HTTP/1.1 200 OK\n"
+                    connstreamout.send(res.encode())
+            except Exception as err:
+                Trace.log(TraceLevel.ERROR, traceback.print_exc())
+        finally:
+            connstreamout.shutdown(socket.SHUT_RDWR)
+            connstreamout.close()
+
+# Class to listen for events and telemetry
```
[the remark this answers is not in the document]
Comment (rakeshsahuu) [reply commands/common/listener_service.py:197]: Done. Moved the Listener thread class and ProcessData sub-thread class to "core/listener.py"

**Review by jskazinski — APPROVED** [review]

**Review by mjenglish — APPROVED** [review]

**Comment by mjenglish** [comment]
@rakeshsahuu Fell free to merge whenever you are ready.

**Pull request merged**